File: examples/dekita/.claude/scripts/validate-hooks-settings.py
#!/usr/bin/env python3
"""settings.json内のフックファイル参照を検証する。

Why:
    削除されたフックへの参照やtypoを検出し、
    実行時エラーを事前に防ぐため。

What:
    - extract_hook_paths(): settings.jsonからフックパスを抽出
    - validate_paths(): ファイル存在を確認

State:
    - reads: .claude/settings.json
    - reads: .claude/hooks/*.py

Remarks:
    - Exit 0: 全参照が有効、Exit 1: 欠落ファイル検出
    - Claude Codeは設定をキャッシュするため、削除後もエラーが継続する問題を防止

Changelog:
    - silenvx/dekita#1300: フック設定検証機能を追加
"""

import json
import re
import sys
from pathlib import Path


def extract_hook_paths(settings: dict, project_dir: Path) -> list[tuple[str, Path]]:
    """Extract all hook file paths from settings.json.

    Args:
        settings: Dictionary with hooks configuration, typically loaded from settings.json.
        project_dir: Path to the project root directory.

    Returns:
        list of (raw_command, resolved_path) tuples.
    """
    hook_paths: list[tuple[str, Path]] = []
    hooks_config = settings.get("hooks", {})

    for hook_type in ["PreToolUse", "PostToolUse", "Stop"]:
        hook_list = hooks_config.get(hook_type, [])
        for hook_group in hook_list:
            for hook in hook_group.get("hooks", []):
                if hook.get("type") == "command":
                    command = hook.get("command", "")
                    # Extract file path from command
                    # Pattern: python3 "$CLAUDE_PROJECT_DIR"/.claude/hooks/xxx.py
                    # or: python3 .claude/hooks/xxx.py
                    # Use non-greedy pattern to stop at .py (not capturing args)
                    match = re.search(
                        r'python3\s+"?\$CLAUDE_PROJECT_DIR"?/([^\s"\']+\.py)', command
                    )
                    if match:
                        relative_path = match.group(1)
                        full_path = project_dir / relative_path
                        hook_paths.append((command, full_path))
                    else:
                        # Fallback: handle direct paths without $CLAUDE_PROJECT_DIR
                        # e.g., "python3 .claude/hooks/xxx.py" or absolute paths
                        # Currently unused but kept for future compatibility
                        match = re.search(r'python3\s+"?([^\s"\']+\.py)', command)
                        if match:
                            path_str = match.group(1)
                            if path_str.startswith("$CLAUDE_PROJECT_DIR"):
                                path_str = path_str.replace("$CLAUDE_PROJECT_DIR", str(project_dir))
                            full_path = Path(path_str)
                            if not full_path.is_absolute():
                                full_path = project_dir / path_str
                            hook_paths.append((command, full_path))

    return hook_paths


def main() -> int:
    # Find project directory (where settings.json is located)
    script_dir = Path(__file__).parent
    project_dir = script_dir.parent.parent  # .claude/scripts -> .claude -> project root

    settings_path = project_dir / ".claude" / "settings.json"

    if not settings_path.exists():
        print(f"⚠️  No settings.json found at {settings_path}")
        return 0  # Not an error - settings might not exist

    try:
        with open(settings_path) as f:
            settings = json.load(f)
    except json.JSONDecodeError as e:
        print(f"❌ Invalid JSON in settings.json: {e}")
        return 1

    hook_paths = extract_hook_paths(settings, project_dir)

    if not hook_paths:
        print("✅ No hook file references found in settings.json")
        return 0

    missing_files: list[tuple[str, Path]] = []
    for command, path in hook_paths:
        if not path.exists():
            missing_files.append((command, path))

    if missing_files:
        print("❌ Missing hook files detected!")
        print("")
        print("The following hook references in settings.json point to non-existent files:")
        print("")
        for command, path in missing_files:
            print(f"  File: {path}")
            print(f"  Command: {command}")
            print("")
        print("To fix:")
        print("  1. Remove the reference from .claude/settings.json, OR")
        print("  2. Create the missing file")
        print("")
        print("Note: After fixing, you may need to restart Claude Code session")
        print("      to clear the cached hook settings.")
        return 1

    print(f"✅ All {len(hook_paths)} hook file references are valid")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: examples/dekita/.claude/scripts/validate_hooks_settings.ts
#!/usr/bin/env bun
/**
 * settings.json内のフックファイル参照を検証する。
 *
 * Why:
 *   削除されたフックへの参照やtypoを検出し、
 *   実行時エラーを事前に防ぐため。
 *
 * What:
 *   - extractHookPaths(): settings.jsonからフックパスを抽出
 *   - validatePaths(): ファイル存在を確認
 *
 * State:
 *   - reads: .claude/settings.json
 *   - reads: .claude/hooks/*.py, .claude/hooks/handlers/*.ts
 *
 * Remarks:
 *   - Exit 0: 全参照が有効、Exit 1: 欠落ファイル検出
 *   - Claude Codeは設定をキャッシュするため、削除後もエラーが継続する問題を防止
 *
 * Changelog:
 *   - silenvx/dekita#1300: フック設定検証機能を追加
 *   - silenvx/dekita#3636: TypeScriptに移植
 */

import { existsSync, readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));

export interface HookConfig {
  type?: string;
  command?: string;
}

export interface HookGroup {
  hooks?: HookConfig[];
}

export interface Settings {
  hooks?: {
    PreToolUse?: HookGroup[];
    PostToolUse?: HookGroup[];
    Stop?: HookGroup[];
    SessionStart?: HookGroup[];
    UserPromptSubmit?: HookGroup[];
  };
}

/**
 * Extract all hook file paths from settings.json.
 */
export function extractHookPaths(settings: Settings, projectDir: string): Array<[string, string]> {
  const hookPaths: Array<[string, string]> = [];
  const hooksConfig = settings.hooks ?? {};

  const hookTypes = [
    "PreToolUse",
    "PostToolUse",
    "Stop",
    "SessionStart",
    "UserPromptSubmit",
  ] as const;

  for (const hookType of hookTypes) {
    const hookList = hooksConfig[hookType] ?? [];
    for (const hookGroup of hookList) {
      for (const hook of hookGroup.hooks ?? []) {
        if (hook.type === "command" && hook.command) {
          const command = hook.command;

          // Pattern: python3 "$CLAUDE_PROJECT_DIR"/.claude/hooks/xxx.py
          // Supports paths with spaces when quoted
          const pythonMatch = command.match(/python3\s+"?\$CLAUDE_PROJECT_DIR"?\/([^"]+\.py)/);
          if (pythonMatch) {
            const relativePath = pythonMatch[1];
            const fullPath = join(projectDir, relativePath);
            hookPaths.push([command, fullPath]);
            continue;
          }

          // Pattern: bun run "$CLAUDE_PROJECT_DIR"/.claude/hooks/handlers/xxx.ts
          // Supports paths with spaces when quoted
          const bunMatch = command.match(/bun\s+run\s+"?\$CLAUDE_PROJECT_DIR"?\/([^"]+\.ts)/);
          if (bunMatch) {
            const relativePath = bunMatch[1];
            const fullPath = join(projectDir, relativePath);
            hookPaths.push([command, fullPath]);
            continue;
          }

          // Fallback: handle direct paths without $CLAUDE_PROJECT_DIR
          // Supports paths with spaces when quoted
          const directMatch = command.match(/(python3|bun\s+run)\s+"?([^"]+\.(py|ts))/);
          if (directMatch) {
            let pathStr = directMatch[2];
            if (pathStr.startsWith("$CLAUDE_PROJECT_DIR")) {
              pathStr = pathStr.replace("$CLAUDE_PROJECT_DIR", projectDir);
            }
            const fullPath = pathStr.startsWith("/") ? pathStr : join(projectDir, pathStr);
            hookPaths.push([command, fullPath]);
          }
        }
      }
    }
  }

  return hookPaths;
}

function main(): number {
  // Find project directory (where settings.json is located)
  const scriptDir = __dirname;
  const projectDir = resolve(scriptDir, "..", ".."); // .claude/scripts -> .claude -> project root

  const settingsPath = join(projectDir, ".claude", "settings.json");

  if (!existsSync(settingsPath)) {
    console.log(`⚠️  No settings.json found at ${settingsPath}`);
    return 0; // Not an error - settings might not exist
  }

  let settings: Settings;
  try {
    const content = readFileSync(settingsPath, "utf-8");
    settings = JSON.parse(content) as Settings;
  } catch (e) {
    if (e instanceof SyntaxError) {
      console.log(`❌ Invalid JSON in settings.json: ${e.message}`);
      return 1;
    }
    throw e;
  }

  const hookPaths = extractHookPaths(settings, projectDir);

  if (hookPaths.length === 0) {
    console.log("✅ No hook file references found in settings.json");
    return 0;
  }

  const missingFiles: Array<[string, string]> = [];
  for (const [command, path] of hookPaths) {
    if (!existsSync(path)) {
      missingFiles.push([command, path]);
    }
  }

  if (missingFiles.length > 0) {
    console.log("❌ Missing hook files detected!");
    console.log("");
    console.log("The following hook references in settings.json point to non-existent files:");
    console.log("");
    for (const [command, path] of missingFiles) {
      console.log(`  File: ${path}`);
      console.log(`  Command: ${command}`);
      console.log("");
    }
    console.log("To fix:");
    console.log("  1. Remove the reference from .claude/settings.json, OR");
    console.log("  2. Create the missing file");
    console.log("");
    console.log("Note: After fixing, you may need to restart Claude Code session");
    console.log("      to clear the cached hook settings.");
    return 1;
  }

  console.log(`✅ All ${hookPaths.length} hook file references are valid`);
  return 0;
}

// Only run main when executed directly (not when imported for testing)
if (import.meta.main) {
  process.exit(main());
}
